presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.optimizations.joins.JoinGraph;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class EliminateCrossJoins
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        if (!SystemSessionProperties.isJoinReorderingEnabled(session)) {
+            return plan;
+        }
+
+        List<JoinGraph> joinGraphs = JoinGraph.buildFrom(plan);
+
+        for (int i = 0; i < joinGraphs.size(); i++) {
+            JoinGraph graph = joinGraphs.get(i);
+            List<Integer> joinOrder = getJoinOrder(graph);
+            if (isOriginalOrder(joinOrder)) {
+                continue;
+            }
+
+            plan = rewriteWith(new Rewriter(idAllocator, graph, joinOrder), plan);
+            joinGraphs = JoinGraph.buildFrom(plan);
```

Comment:
I remembered what was the issue. `joinOrder` is meaningless without `joinGraph`, because you don't know what plan nodes is it referring to. On the other hand, each rewrite invalidates join graphs that are above rewritten part of the plan. That's why I had to rebuild it after every rewrite.

I have changed this code, so that it rewrites joins from top to bottom. That should allow me to avoid rebuilding join graphs after every rewrite.